mmv1/third_party/terraform/services/storage/resource_storage_folder_test.go

```diff
@@ -0,0 +1,238 @@
+		Steps: []resource.TestStep{
+			{
+				Config: testAccStorageFolder_storageBucket(bucketName, true, true) + testAccStorageFolder_storageFolder(true),
+			},
+			{
+				ResourceName:            "google_storage_folder.folder",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"bucket", "recursive", "force_destroy"},
+			},
+		},
+	})
+}
+
+func TestAccStorageFolder_hnsDisabled(t *testing.T) {
+	t.Parallel()
+
+	bucketName := acctest.TestBucketName(t)
+
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccStorageFolder_storageBucket(bucketName, false, true) + testAccStorageFolder_storageFolder(true),
+				ExpectError: regexp.MustCompile("Error creating Folder: googleapi: Error 409: The bucket does not support hierarchical namespace., conflict"),
+			},
+		},
+	})
+}
+
+func TestAccStorageFolder_FolderForceDestroy(t *testing.T) {
+	t.Parallel()
+
+	bucketName := acctest.TestBucketName(t)
+
+	data := []byte("data data data")
+
+	testFile := getNewTmpTestFile(t, "tf-test")
+	if err := ioutil.WriteFile(testFile.Name(), data, 0644); err != nil {
+		t.Errorf("error writing file: %v", err)
+	}
+
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccStorageFolder_storageBucketObject(bucketName, true, true, testFile.Name()),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckStorageBucketUploadItem(t, bucketName),
+				),
+			},
+		},
+	})
+}
+
+func TestAccStorageFolder_DeleteEmptyFolderWithForceDestroyDefault(t *testing.T) {
+	t.Parallel()
+
+	bucketName := acctest.TestBucketName(t)
+
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccStorageFolder_storageBucket(bucketName, true, true) + testAccStorageFolder_storageOneFolder(false),
+			},
+		},
+	})
+}
+
+func TestAccStorageFolder_FailDeleteNonEmptyFolder(t *testing.T) {
+	t.Parallel()
+
+	bucketName := acctest.TestBucketName(t)
+	folderName := "folder/"
+
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccStorageFolder_storageBucket(bucketName, true, true) + testAccStorageFolder_storageOneFolder(false),
+				Check: resource.ComposeTestCheckFunc(
+					testAccStorageCreatSubFolder(t, bucketName, folderName),
+					testAccStorageDeleteFolder(t, bucketName, folderName),
```

Comment:
I'm not sure I'm understanding this part. Why are we deleting `folder/` here? Shouldn't we instead do that with Terraform config (by removing it)?

Reply:
here we are creating a bucket and a folder with force_destroy false in the config step and later created a sub-folder inside a folder and now tried to delete parent folder, as force_destory set to false on parent folder we expect an error as parent folder is not empty (sub-folder exists),  seems this is the best way to produce this test case